packages/react-select/src/Select.js

```diff
@@ -363,29 +487,49 @@ export default class Select extends Component<Props, State> {
   constructor(props: Props) {
     super(props);
     const { value } = props;
-    this.cacheComponents = memoizeOne(this.cacheComponents, isEqual).bind(this);
-    this.cacheComponents(props.components);
     this.instancePrefix =
       'react-select-' + (this.props.instanceId || ++instanceId);
+    this.state.selectValue = cleanValue(value);
+  }
+  static getDerivedStateFromProps(props: Props, state: State) {
+    const { prevProps } = state;
+    const { options, value, menuIsOpen, inputValue } = props;
+    let newMenuOptionsState = {};
+    if (
+      prevProps &&
+      (value !== prevProps.value ||
+        options !== prevProps.options ||
+        menuIsOpen !== prevProps.menuIsOpen ||
+        inputValue !== prevProps.inputValue)
+    ) {
+      const selectValue = cleanValue(value);
+      const focusableOptions = menuIsOpen
+        ? getFocusableOptions(props, state, selectValue)
+        : [];
+      const focusedValue = getNextFocusedValue(state, selectValue);
+      const focusedOption = getNextFocusedOption(state, focusableOptions);
+      newMenuOptionsState = { selectValue, focusedOption, focusedValue };
+    }
+    // some updates should toggle the state of the input visibility
+    const { inputIsHiddenAfterUpdate } = state;
+    const newInputIsHiddenState =
+      inputIsHiddenAfterUpdate != null && !isEqual(props, state.prevProps)
+        ? {
+            inputIsHidden: inputIsHiddenAfterUpdate,
+            inputIsHiddenAfterUpdate: undefined,
+          }
+        : {};
+    const newClearFocusValueOnUpdateState =
+      state.clearFocusValueOnUpdate && !isEqual(props, state.prevProps)
+        ? { clearFocusValueOnUpdate: false }
+        : {};
 
-    const selectValue = cleanValue(value);
-
-    this.buildMenuOptions = memoizeOne(
-      this.buildMenuOptions,
-      (newArgs: any, lastArgs: any) => {
-        const [newProps, newSelectValue] = (newArgs: [Props, OptionsType]);
-        const [lastProps, lastSelectValue] = (lastArgs: [Props, OptionsType]);
-
-        return isEqual(newSelectValue, lastSelectValue)
-          && isEqual(newProps.inputValue, lastProps.inputValue)
-          && isEqual(newProps.options, lastProps.options);
-      }).bind(this);
```

Comment:
I removed the memoization here since `buildMenuOptions` is refactored in this PR in a way that makes it hard to memoize it the same way it was before. The memoization was originally added in https://github.com/JedWatson/react-select/pull/3868, but it doesn't look the memoization was an essential part of that PR.